### PR TITLE
Add useBuiltIns and useESModules options to transform-runtime

### PR DESCRIPTION
**Pull request by Kovensky**
useBuiltIns uses versions of the helpers that do not import even
internal polyfills from core-js.

useESModules uses versions of the helpers that do not go through
transform-es2015-modules-commonjs. This allows for smaller builds in
module systems like webpack, as it doesn't need to preserve commonjs
semantics.

This includes changes to the babel-runtime build-dist script, which
will build the versions of the runtime helpers to be used by
combinations of useBuiltIns and useESModules.

This is untested other than by looking at the output and seeing if it makes sense >_<

`{ polyfill: false, useBuiltIns: true, useESModules: true }` will be perfect for webpack/rollup
users that want to deduplicate helpers without accidentally bundling an extra
core-js when babel-polyfill is already in use, or when, say, targeting electron.

### CODE REVIEW

**Comment by hzoo** [comment]
Maybe we should actually add the generated files to git so we can actually compare what's happening.. Don't know how we can write tests

**Comment by hzoo** [comment]
Yeah it would be nice if we could diff previous code or diff the options somehow

**Comment by xtuc** [comment]
Could we add it as a snapshot in the tests?

**Comment by Kovensky** [comment]
I tried committing the files to git, but it would end up creating a huge commit adding 300+ files, and at least 4 of them would need to be updated every single time a helper got changed :cry:

I added fixtures for the transform-runtime output, but no idea how babel-runtime could be tested. Any exec test that would pass with transform-runtime would also pass without using transform-runtime at all, and it's not possible to exec-test the useESModules output.

File: packages/babel-runtime/scripts/build-dist.js [hunk 1]
```diff
@@ -46,29 +46,40 @@ function writeFile(filename, content) {
...
+}
 
 function buildRuntimeRewritePlugin(relativePath, helperName) {
   return {
     pre: function (file) {
       var original = file.get("helperGenerator");
       file.set("helperGenerator", function(name) {
         // make sure that helpers won't insert circular references to themselves
-        if (name === helperName) return;
+        if (name === helperName) return false;
```
Comment (loganfsmyth) [inline packages/babel-runtime/scripts/build-dist.js]: Why `false`?

File: packages/babel-runtime/scripts/build-dist.js [hunk 2]
```diff
@@ -82,23 +93,40 @@ function buildRuntimeRewritePlugin(relativePath, helperName) {
...
+    for (const helperName of helpers.list) {
+      writeFile(`${dirname}${helperName}.js`, buildHelper(helperName, modules, builtin));
+
+      // compat
+      var helperAlias = kebabCase(helperName);
+      var content = !modules
+        ? `export { default } from \"./${helperName}.js\";`
+        : "module.exports = require(\"./" + helperName + ".js\");";
+      writeFile(`${dirname}_${helperAlias}.js`, content);
+      if (helperAlias !== helperName) writeFile(`${dirname}${helperAlias}.js`, content);
```
Comment (loganfsmyth) [inline packages/babel-runtime/scripts/build-dist.js]: We should really kill all these aliases in 7.x. If you're up for it, maybe a PR another this one? Let's keep this separate so if we want to backport it is easier.

**Comment by loganfsmyth** [comment]
This looks awesome to me. The only case that comes to mind is that it's unfortunate people using `useBuiltins` still have to depend on `core-js` even though they aren't using it.

Let's land this as-is, but it might be good for us to think about if we want to explore making a separate module for "only helpers" and one for "helpers with polyfill" or something.

In the long run I'd love to have the helpers currently living in `babel-helpers` instead live inside their respective plugins, which would allow us to repurpose `babel-helpers`. It's a clear name for something like that.

**Comment by hzoo** [comment]
preset-env's useBuiltIns option outputs core-js which isn't very intuitive so it would be nice for polyfill to just have the modules. Would the transform easier too cc @zloirock

**Comment by Kovensky** [comment]
Yeah, I took the `useBuiltIns` name from preset-env.

@loganfsmyth the point of `useBuiltIns` is specifically to not depend on core-js and just use whatever is in the environment, but I wouldn't be surprised if some helpers only work with either native or core-js's implementation.

**Pull request merged**